src/lib/flow/sol-flow-composed.c

```diff
@@ -0,0 +1,530 @@
+        return SOL_FLOW_PACKET_TYPE_STRING;
+    if (!strcasecmp(type, "boolean"))
+        return SOL_FLOW_PACKET_TYPE_BOOLEAN;
+    if (!strcasecmp(type, "byte"))
+        return SOL_FLOW_PACKET_TYPE_BYTE;
+    if (!strcasecmp(type, "blob"))
+        return SOL_FLOW_PACKET_TYPE_BLOB;
+    if (strcasecmp(type, "rgb"))
+        return SOL_FLOW_PACKET_TYPE_RGB;
+    if (strcasecmp(type, "location"))
+        return SOL_FLOW_PACKET_TYPE_LOCATION;
+    if (strcasecmp(type, "timestamp"))
+        return SOL_FLOW_PACKET_TYPE_TIMESTAMP;
+    if (strcasecmp(type, "direction-vector"))
+        return SOL_FLOW_PACKET_TYPE_DIRECTION_VECTOR;
+    if (strcasecmp(type, "error"))
+        return SOL_FLOW_PACKET_TYPE_ERROR;
+    return NULL;
+}
+
+static int
+simple_port_process(struct sol_flow_node *node, void *data, uint16_t port,
+    uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    struct composed_node_data *cdata = data;
+    struct sol_flow_packet *composed;
+    uint16_t i;
+
+    if (cdata->inputs[port]) {
+        sol_flow_packet_del(cdata->inputs[port]);
+        cdata->inputs[port] = NULL;
+    }
+
+    cdata->inputs[port] = sol_flow_packet_dup(packet);
+    SOL_NULL_CHECK(cdata->inputs[port], -ENOMEM);
+
+    for (i = 0; i < cdata->inputs_len; i++) {
+        if (!cdata->inputs[i])
+            break;
+    }
+
+    if (i != cdata->inputs_len)
+        return 0;
+
+    composed = sol_flow_packet_new(cdata->composed_type, cdata->inputs);
+    SOL_NULL_CHECK(composed, -ENOMEM);
+    return sol_flow_send_packet(node, 0, composed);
+}
+
+static int
+composed_port_process(struct sol_flow_node *node, void *data, uint16_t port,
+    uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    int r;
+    uint16_t len, i;
+    struct sol_flow_packet **children, *out_packet;
+
+    r = sol_flow_packet_get_composed_members_len(
+        sol_flow_packet_get_type(packet), &len);
+    SOL_INT_CHECK(r, < 0, r);
+
+    children = alloca(len * sizeof(struct sol_flow_packet *));
+
+    r = sol_flow_packet_get(packet, children);
+    SOL_INT_CHECK(r, < 0, r);
+
+    for (i = 0; i < len; i++) {
+        out_packet = sol_flow_packet_dup(children[i]);
+        SOL_NULL_CHECK(out_packet, -ENOMEM);
+        r = sol_flow_send_packet(node, i, out_packet);
+        SOL_INT_CHECK(r, < 0, r);
+    }
+
+    return 0;
+}
+
+static int
+setup_simple_ports(struct sol_vector *in_ports, const struct sol_str_slice contents, bool is_input)
+{
+    struct sol_vector tokens;
+    struct sol_str_slice *slice;
+    struct composed_node_port_type *port_type;
+    const struct sol_flow_packet_type *packet_type;
+    struct sol_buffer buf;
+    char *token, *name, *type;
+    size_t i_slice;
+    uint16_t i;
+    int r;
+
+    sol_buffer_init(&buf);
+    for (i_slice = 0; i_slice < contents.len; i_slice++) {
+        if (isspace(contents.data[i_slice]))
+            continue;
+        sol_buffer_append_slice(&buf, SOL_STR_SLICE_STR(contents.data + i_slice, 1));
```

Comment:
we now have a `append_char()` that would make this simpler, but please rewrite this block so we're optimal: scan for space and then copy the previous slice if a space is found or if end-of-slice. If space, then keep increasing the "pendings_slice" start, once a non-space is found, you set its data and start increasing the length.